packages/event_store/src/httpServer.js

```diff
@@ -358,21 +358,92 @@ let httpServer = (models) => {
+        for (const genesisAccount of genesisAccounts) {
+            indexedGenesisBalances[genesisAccount.accountHash] = BigNumber.from(genesisAccount.balance);
+        }
+
+        // Rebalance genesis account balances based on the transfers between genesis accounts
+        const internalGenesisAccountTransfers = await storage.getTokensMovedBetweenGenesisAccounts();
+        for (const transfer of internalGenesisAccountTransfers) {
+            // Note, that here we can also count a transfer of non-genesis tokens (since they are fungible)
+            // It is addressed in the transfer overflow correction logic
+            indexedGenesisBalances[transfer.fromAccount] = indexedGenesisBalances[transfer.fromAccount]
+                .sub(transfer.amount);
+
+            indexedGenesisBalances[transfer.toAccount] = indexedGenesisBalances[transfer.toAccount]
+                .add(transfer.amount);
+        }
+
+        // If a genesis account transferred to other genesis accounts more tokens than it had
+        // in the balance on genesis then we should adjust the amount of the genesis tokens that
+        // became circulating by that overflow value
+        let internalTransfersOverflowCorrection = BigNumber.from(0);
+        for (const transfer of internalGenesisAccountTransfers) {
+            if (indexedGenesisBalances[transfer.fromAccount].lte(0)) {
+                internalTransfersOverflowCorrection = internalTransfersOverflowCorrection
+                    .add(indexedGenesisBalances[transfer.fromAccount]);
+            }
+        }
+
+        // Calculate genesis tokens transferred out of the genesis accounts
+        let releasedGenesisMotesAmount = BigNumber.from(0);
+
+        const externalGenesisAccountTransfers = await storage.getTokensMovedOutOfGenesisAccounts();
+        for (const transfer of externalGenesisAccountTransfers) {
+            if (indexedGenesisBalances[transfer.fromAccount].lte(0)) {
+                // If no genesis tokens left in the account then do nothing
+                continue;
+            }
+
+            if (indexedGenesisBalances[transfer.fromAccount].gte(transfer.amount)) {
+                // If genesis token amount is greater than the transfer amount then
+                // increase the circulating genesis tokens by this amount
+                releasedGenesisMotesAmount = releasedGenesisMotesAmount
+                    .add(transfer.amount);
+
+                indexedGenesisBalances[transfer.fromAccount] = indexedGenesisBalances[transfer.fromAccount]
+                    .sub(transfer.amount);
+            }
+            else {
+                // Otherwise increase the circulating genesis tokens by the account balance remainder
+                releasedGenesisMotesAmount = releasedGenesisMotesAmount
+                    .add(indexedGenesisBalances[transfer.fromAccount]);
+
+                indexedGenesisBalances[transfer.fromAccount] = indexedGenesisBalances[transfer.fromAccount]
```

Comment:
@ihor doesn't this always set it `indexedGenesisBalances[transfer.fromAccount]` to `0`?  You subtracting a value of itself?

Reply:
Yes, it does. That if-else statement is deducting Math.min(transferAmount, remainingBalance). I wanted to preserve this structure, and that's why I'm subtracting the value itself instead of setting it to 0.